File: Modules/CalculateMath.m
function handles = CalculateMath(handles)

% Help for the Calculate Math module:
% Category: Measurement
%
% SHORT DESCRIPTION:
% This module can take measurements produced by previous modules and
% performs basic arithmetic operations.
% *************************************************************************
%
% The arithmetic operations available in this module include addition,
% subtraction, multiplication and division. The operation can be chosen
% by adjusting the operations setting. The resulting data can also be
% logged or raised to a power. This data can then be used in other
% calculations and can be used in Classify Objects.
%
% This module currently works on an object-by-object basis (it calculates
% the requested operation for each object) but can also apply the operation
% for measurements made for entire images.
%
% Feature Number:
% The feature number specifies which features from the Measure module(s)
% will be used for the operation. See each Measure module's help for the
% numbered list of the features measured by that module.
%
% Saving:
% The math measurements are stored as 'Math_...'. If both measures are 
% image-based, then a single calculation (per cycle) will be stored as 'Image' data.  
% If one measure is object-based and one image-based, then the calculations will
% be stored associated with the object, one calculation per object.  If both are 
% objects, then the calculations are stored with both objects.
%
% Category: 'Math'
% Features measured:                         Feature Number:
% (As named in module's last setting)     |       1
%
% Note: If you want to use the output of this module in a subsequesnt
% calculation, we suggest you specify the output name rather than use
% Automatic naming.
%
% See also CalculateRatios, all Measure modules.

% CellProfiler is distributed under the GNU General Public License.
% See the accompanying file LICENSE for details.
%
% Developed by the Whitehead Institute for Biomedical Research.
% Copyright 2003,2004,2005.
%
% Please see the AUTHORS file for credits.
%
% Website: http://www.cellprofiler.org
%
% $Revision$

% MBray 2009_03_20: Comments on variables for pyCP upgrade
%
% Recommended variable order (setting, followed by current variable in MATLAB CP)
% (1) What do you want to call the measurement calculated by this module? (OutputFeatureName)
% (2) What operation would you like to perform? (Operation)
% (do we currently allow 'raise to a power' here?)
% Perhaps show settings for the operands in two columns (eg, numerator settings on
% left, denominator settings on right). Maybe show mathematical symbol (based on answer
% to (2)) between the columns. Or, if this is too time-consuming, at least
% show, based on the selection in (2), which operand is which. In other
% words, if they choose Division, then show: Operand 1/operand 2. If they
% choose subtraction, show: operand 1 - operand 2.
%
% (2.5) What type of number is the first operand? 
% CHOICES: (a) measurement of an individual object, (b) measurement from a whole
% image, (c) number
% If they choose (a) then provide the list of objects, if (b) then list of
% images, if (c) then provide an edit box for them to enter something. This
% new ability to calculate per-image features or enter numbers then affects
% the rest of the questions below (they are currently written for
% per-object measurements, but hopefully it's clear how to adjust them for
% per-image measurements or for a number the user specifies):
%
% (3a) Which object's measurement would you like to use as the first operand? (ObjectName{1})
% (3b) What is the measurement category? (Category{1})
% (3c) What is the measurement feature? (FeatureNumber{1})
% (3d) (If the answer to (3c) involves a scale) What scale was used to 
%      calculate the feature? (SizeScale{1}) 
%      (If the answer to (3c) involves an image) What image was used to
%      calculate the feature? (ImageName{1})
% (3e) What number would you like to multiply the operand by?  (MultiplyFactor1)
% (3f) What power would you like to raise the operand to? (this feature
% currently doesn't exist; should we ask the user what operation they want
% to perform on the first operand prior to calculating with the second
% operand, then offer multiply, raise to a power, and perhaps other
% options? or is this just getting way too confusing?)
%
%
% (3.5) Again, insert "what type of number is the second operand?"
%
% (4a) Which object's measurement would you like to use as the second operand? (ObjectName{2})
% (4b) What is the measurement category? (Category{2})
% (4c) What is the measurement feature? (FeatureNumber{2})
% (4d) (If the answer to (4c) involves a scale) What scale was used to 
%      calculate the feature? (SizeScale{2}) 
%      (If the answer to (4c) involves an image) What image was used to
%      calculate the feature? (ImageName{2})
% (4e) What number would you like to multiply the operand by? (MultiplyFactor2)
% (4f) What power would you like to raise the operand to? (this feature
% currently doesn't exist)
%
% question: should we allow the user to choose additional operands (perhaps just for
% addition and multiplication) or just let them use an additional
% CalculateMath module in the pipeline for that sort of thing? I think the
% latter would be fine, especially if it keeps this module from ending up
% being confusing.
%
% Note: we could eliminate the next two questions because the user could
% just put an additional CalculateMath module in the pipeline, right?
% (5a) What power would you like to raise the result to? (Power)
% (5b) What number would you like to multiply the above result by? (MultiplyFactor3)
%
% NOTES:
% (i) Setting (1) should not have an automatic option; the user should
% specify the name explicitly. (Anne says: why not, Mark? Do we tend to run
% into trouble with the name being too long? If so, then this makes sense.
% We should make sure then, that the custom-entered feature name is
% choosable in other drop down menus downstream in the pipeline - I think
% we have a 'custom' measurement type.)
% (ii) Measurement category/feature/image/scale settings should only be shown if
% the measurement hierarchy requires it.
% (iii) Some clarification will be needed to show the order of operations to
% the user

%%%%%%%%%%%%%%%%%
%%% VARIABLES %%%
%%%%%%%%%%%%%%%%%

drawnow

[CurrentModule, CurrentModuleNum, ModuleName] = CPwhichmodule(handles);

%textVAR01 = Which object would you like to use as the first measurement?
%choiceVAR01 = Image
%infotypeVAR01 = objectgroup
%inputtypeVAR01 = popupmenu
ObjectName{1} = char(handles.Settings.VariableValues{CurrentModuleNum,1});

%textVAR02 = Which category of measurements would you like to use?
%inputtypeVAR02 = popupmenu category
Category{1} = char(handles.Settings.VariableValues{CurrentModuleNum,2});

%textVAR03 = Which feature do you want to use? (Enter the feature number or name - see help for details)
%defaultVAR03 = 1
%inputtypeVAR03 = popupmenu measurement
FeatureNumber{1} = handles.Settings.VariableValues{CurrentModuleNum,3};

%textVAR04 = For INTENSITY, AREAOCCUPIED or TEXTURE features, which image's measurements would you like to use?
%infotypeVAR04 = imagegroup
%inputtypeVAR04 = popupmenu
ImageName{1} = char(handles.Settings.VariableValues{CurrentModuleNum,4});

%textVAR05 = For TEXTURE, RADIAL DISTRIBUTION, OR NEIGHBORS features, what previously measured size scale (TEXTURE OR NEIGHBORS) or previously used number of bins (RADIALDISTRIBUTION) do you want to use?
%defaultVAR05 = 1
%inputtypeVAR05 = popupmenu scale
SizeScale{1} = str2double(handles.Settings.VariableValues{CurrentModuleNum,5});

%textVAR06 = Which object would you like to use as the second measurement?
%choiceVAR06 = Image
%infotypeVAR06 = objectgroup
%inputtypeVAR06 = popupmenu
ObjectName{2} = char(handles.Settings.VariableValues{CurrentModuleNum,6});

%textVAR07 = Which category of measurements would you like to use?
%inputtypeVAR07 = popupmenu category
Category{2} = char(handles.Settings.VariableValues{CurrentModuleNum,7});

%textVAR08 = Which feature do you want to use? (Enter the feature number or name - see help for details)
%defaultVAR08 = 1
%inputtypeVAR08 = popupmenu measurement
FeatureNumber{2} = handles.Settings.VariableValues{CurrentModuleNum,8};

%textVAR09 = For INTENSITY, AREAOCCUPIED or TEXTURE features, which image's measurements would you like to use?
%infotypeVAR09 = imagegroup
%inputtypeVAR09 = popupmenu
ImageName{2} = char(handles.Settings.VariableValues{CurrentModuleNum,9});

%textVAR10 = For TEXTURE, RADIAL DISTRIBUTION, OR NEIGHBORS features, what previously measured size scale (TEXTURE OR NEIGHBORS) or previously used number of bins (RADIALDISTRIBUTION) do you want to use?
%defaultVAR10 = 1
%inputtypeVAR10 = popupmenu scale
SizeScale{2} = str2double(handles.Settings.VariableValues{CurrentModuleNum,10});

%textVAR11 = Do you want the log (base 10) of the ratio?
%choiceVAR11 = No
%choiceVAR11 = Yes
%inputtypeVAR11 = popupmenu
LogChoice = char(handles.Settings.VariableValues{CurrentModuleNum,11});

%textVAR12 = Enter a factor to multiply the first feature by (before other operations):
%defaultVAR12 = 1
MultiplyFactor1 = str2double(handles.Settings.VariableValues{CurrentModuleNum,12});

%textVAR13 = Enter a factor to multiply the second feature by (before other operation):
%defaultVAR13 = 1
MultiplyFactor2 = str2double(handles.Settings.VariableValues{CurrentModuleNum,13});

%textVAR14 = Enter an exponent to raise the result to (*after* chosen operation below)?
%defaultVAR14 = 1
Power = str2double(handles.Settings.VariableValues{CurrentModuleNum,14});

%textVAR15 = Enter a factor to multiply the result by (*after* chosen operation below)?
%defaultVAR15 = 1
MultiplyFactor3 = str2double(handles.Settings.VariableValues{CurrentModuleNum,15});

%textVAR16 = Operation?
%choiceVAR16 = Multiply
%choiceVAR16 = Divide
%choiceVAR16 = Add
%choiceVAR16 = Subtract
%inputtypeVAR16 = popupmenu
Operation = char(handles.Settings.VariableValues(CurrentModuleNum, 16));

%textVAR17 = What do you want to call the output calculated by this module? The prefix, "Math_" will be applied to your entry or simply leave as "Automatic" and a sensible name will be generated.'
%defaultVAR17 = Automatic
OutputFeatureName = char(handles.Settings.VariableValues(CurrentModuleNum,17));
%%%VariableRevisionNumber = 6

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% MAKE MEASUREMENTS & SAVE TO HANDLES STRUCTURE %%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
drawnow

SetBeingAnalyzed = handles.Current.SetBeingAnalyzed;

for idx = 1:2
    try
        FeatureName{idx} = CPgetfeaturenamesfromnumbers(handles, ObjectName{idx}, ...
            Category{idx}, FeatureNumber{idx}, ImageName{idx},SizeScale{idx});

        Measurements{idx} = handles.Measurements.(ObjectName{idx}).(FeatureName{idx}){SetBeingAnalyzed};
    catch
        error([lasterr '  Image processing was canceled in the ', ModuleName, ...
            ' module (#' num2str(CurrentModuleNum) ...
            ') because an error ocurred when retrieving the data.  '...
            'Likely the category of measurement you chose, ',...
            Category{idx}, ', was not available for ', ...
            ObjectName{idx},' with feature number ' FeatureNumber{idx} ...
            ', possibly specific to image ''' ImageName{idx} ''' and/or ' ...
            'Texture Scale = ' num2str(SizeScale{idx}) '.']);
    end
end

%%% Check to make sure multiply factors are valid entries. If not change to
%%% default and warn user.
if isnan(MultiplyFactor1)
    if isempty(findobj('Tag',['Msgbox_' ModuleName ', ModuleNumber ' num2str(CurrentModuleNum) ': First multiply factor invalid']))
        CPwarndlg(['The first image multiply factor you have entered in the ', ModuleName, ' module is invalid, it is being reset to 1.'],[ModuleName ', ModuleNumber ' num2str(CurrentModuleNum) ': First multiply factor invalid'],'replace');
    end
    MultiplyFactor1 = 1;
end
if isnan(MultiplyFactor2)
    if isempty(findobj('Tag',['Msgbox_' ModuleName ', ModuleNumber ' num2str(CurrentModuleNum) ': Second multiply factor invalid']))
        CPwarndlg(['The second image multiply factor you have entered in the ', ModuleName, ' module is invalid, it is being reset to 1.'],[ModuleName ', ModuleNumber ' num2str(CurrentModuleNum) ': Second multiply factor invalid'],'replace');
    end
    MultiplyFactor2 = 1;
end
    
% Check sizes (note, 'Image' measurements have length=1)
if length(Measurements{1}) ~= length(Measurements{2}) && ...
        ~(length(Measurements{1}) ==1 || length(Measurements{2}) == 1)
    error(['Image processing was canceled in the ', ModuleName, ' module because the specified object names ',ObjectName{1},' and ',ObjectName{2},' do not have the same object count.']);
end

% Construct field name
if isempty(OutputFeatureName) || strcmp(OutputFeatureName,'Automatic') == 1
    FullFeatureName = CPjoinstrings('Math',ObjectName{1},FeatureName{1},...
                                    Operation,ObjectName{2},FeatureName{2});

    % Since Matlab's max name length is 63, we need to truncate the fieldname
    TruncFeatureName = CPtruncatefeaturename(FullFeatureName);
else
    TruncFeatureName = CPjoinstrings('Math',OutputFeatureName);
end
% Do Math
if( strcmpi(Operation, 'Multiply') )
    FinalMeasurements = (MultiplyFactor1.*Measurements{1}) .* (MultiplyFactor2.*Measurements{2});
elseif( strcmpi(Operation, 'Divide') )
    Measurements{2}(Measurements{2}==0) = NaN;
    Measurements{2}(isnan(Measurements{2})) = CPnanmean(Measurements{2});
    
    if ~all(isnan(Measurements{2}))
    FinalMeasurements = (MultiplyFactor1.*Measurements{1}) ./ (MultiplyFactor2.*Measurements{2});
    else
    FinalMeasurements = NaN;
    CPwarndlg(['A ratio of ' NumerDenomMeasurements{1} ' and ' NumerDenomMeasurements{2} ...
        ' within ' ModuleName ' on cycle '  str2double(SetBeingAnalyzed) ...
        ' resulted in all NaNs.  You may want to check your settings.'])
    end

elseif( strcmpi(Operation, 'Add') )
    FinalMeasurements = (MultiplyFactor1.*Measurements{1}) + (MultiplyFactor2.*Measurements{2});
elseif( strcmpi(Operation, 'Subtract') )
    FinalMeasurements = (MultiplyFactor1.*Measurements{1}) - (MultiplyFactor2.*Measurements{2});
end
    
if strcmp(LogChoice,'Yes')
    FinalMeasurements = log10(FinalMeasurements);
end

if ~isnan(Power)
    FinalMeasurements = FinalMeasurements .^ Power;
end

if ~isnan(MultiplyFactor3)
    FinalMeasurements = FinalMeasurements.*MultiplyFactor3;
end

% Save, depending on type of measurement (ObjectName)
% Note that Image measurements are scalars, while Objects are potentially vectors
if strcmp(ObjectName{1}, 'Image') && strcmp(ObjectName{2}, 'Image'),
    handles = CPaddmeasurements(handles,'Image',TruncFeatureName,FinalMeasurements);
else
    if ~strcmp(ObjectName{1}, 'Image'),
        handles = CPaddmeasurements(handles,ObjectName{1},TruncFeatureName,FinalMeasurements);
    end
    if ~strcmp(ObjectName{1}, ObjectName{2}) && ~ strcmp(ObjectName{2},'Image')
        handles = CPaddmeasurements(handles,ObjectName{2},TruncFeatureName,FinalMeasurements);
    end
end

%%%%%%%%%%%%%%%%%%%%%%%
%%% DISPLAY RESULTS %%%
%%%%%%%%%%%%%%%%%%%%%%%
drawnow

% TODO add display of numbers, as in CalculateRatios, and remove CPclosefigure

%%% The figure window display is unnecessary for this module, so it is
%%% closed during the starting image cycle.
CPclosefigure(handles,CurrentModule)

File: pyCellProfiler/cellprofiler/modules/calculatemath.py
'''calculatemath.py - the CalculateMath module

CellProfiler is distributed under the GNU General Public License.
See the accompanying file LICENSE for details.

Developed by the Broad Institute
Copyright 2003-2009

Please see the AUTHORS file for credits.

Website: http://www.cellprofiler.org
'''
__version__="$Revision$"

import numpy as np

import cellprofiler.cpmodule as cpm
import cellprofiler.measurements as cpmeas
import cellprofiler.settings as cps

O_MULTIPLY = "Multiply"
O_DIVIDE = "Divide"
O_ADD = "Add"
O_SUBTRACT = "Subtract"

O_ALL = [O_MULTIPLY, O_DIVIDE, O_ADD, O_SUBTRACT]

MC_IMAGE = cpmeas.IMAGE
MC_OBJECT = "Object"
MC_ALL = [MC_IMAGE, MC_OBJECT]

class CalculateMath(cpm.CPModule):
    '''SHORT DESCRIPTION:
This module can take measurements produced by previous modules and
performs basic arithmetic operations.
*************************************************************************

The arithmetic operations available in this module include addition,
subtraction, multiplication and division. The operation can be chosen
by adjusting the operations setting. The resulting data can also be
logged or raised to a power. This data can then be used in other
calculations and can be used in Classify Objects.

This module currently works on an object-by-object basis (it calculates
the requested operation for each object) but can also apply the operation
for measurements made for entire images.

Saving:
The math measurements are stored as 'Math_...'. If both measures are 
image-based, then a single calculation (per cycle) will be stored as 'Image' data.  
If one measure is object-based and one image-based, then the calculations will
be stored associated with the object, one calculation per object.  If both are 
objects, then the calculations are stored with both objects.

Note: If you want to use the output of this module in a subsequesnt
calculation, we suggest you specify the output name rather than use
Automatic naming.

See also CalculateRatios, all Measure modules.
'''

    module_name = "CalculateMath"
    category="Measurement"
    variable_revision_number = 1
    
    def create_settings(self):
        class Operand(object):
            '''Represents the collection of settings needed by each operand'''
            def __init__(self, index, operation):
                self.__index = index
                self.__operation = operation
                self.__operand_choice = cps.Choice(self.operand_choice_text(), MC_ALL)
                self.__operand_objects = cps.ObjectNameSubscriber(self.operand_objects_text(),"None")
                self.__operand_measurement = cps.Measurement(self.operand_measurement_text(),
                                                             self.object_fn)
                self.__multiplicand = cps.Float("What number would you like to multiply the above operand by?",1)
                self.__exponent = cps.Float("What power would you like to raise the above operand to?",1)
            
            @property
            def operand_choice(self):
                '''Either MC_IMAGE for image measurements or MC_OBJECT for object'''
                return self.__operand_choice
            
            @property
            def operand_objects(self):
                '''Get measurements from these objects'''
                return self.__operand_objects
            
            @property
            def operand_measurement(self):
                '''The measurement providing the value of the operand'''
                return self.__operand_measurement
            
            @property 
            def multiplicand(self):
                '''Premultiply the measurement by this value'''
                return self.__multiplicand
            
            @property
            def exponent(self):
                '''Raise the measurement to this power'''
                return self.__exponent
            
            @property
            def object(self):
                '''The name of the object for measurement or "Image"'''
                if self.operand_choice == MC_IMAGE:
                    return cpmeas.IMAGE
                else:
                    return self.operand_objects.value
                
            def object_fn(self):
                if self.__operand_choice == MC_IMAGE:
                    return cpmeas.IMAGE
                elif self.__operand_choice == MC_OBJECT:
                    return self.__operand_objects.value
                else:
                    raise NotImplementedError("Measurement type %s is not supported"%
                                              self.__operand_choice.value)
            def operand_name(self):
                '''A fancy name based on what operation is being performed'''
                if self.__index == 0:
                    return ("first operand" 
                            if self.__operation in (O_ADD, O_MULTIPLY) else
                            "minuend" if self.__operation == O_SUBTRACT else
                            "numerator")
                elif self.__index == 1:
                    return ("second operand" 
                            if self.__operation in (O_ADD, O_MULTIPLY) else
                            "subtrahend" if self.__operation == O_SUBTRACT  else 
                            "denominator")
            
            def operand_choice_text(self):
                return self.operand_text("Is the %s an image or object measurement?") 
            
            def operand_objects_text(self):
                return self.operand_text("Which objects do you want to measure for the %s?")
            
            def operand_text(self, format):
                return format % self.operand_name()
                
            def operand_measurement_text(self): 
                return self.operand_text("What measurement do you want to use for the %s?")
            
            def settings(self):
                '''The operand settings to be saved in the output file'''
                return [self.operand_choice, self.operand_objects, 
                        self.operand_measurement, self.multiplicand, self.exponent]
            
            def visible_settings(self):
                '''The operand settings to be displayed'''
                self.operand_choice.text = self.operand_choice_text()
                self.operand_objects.text = self.operand_objects_text()
                self.operand_measurement.text = self.operand_measurement_text()
                return ([self.operand_choice] +
                        ([self.operand_objects] if self.operand_choice == MC_OBJECT
                          else []) +
                        [self.operand_measurement,self.multiplicand, self.exponent])
            
        self.output_feature_name = cps.Text("What do you want to call the measurement calculated by this module?",
                                            "Measurement")
        self.operation = cps.Choice("What operation would you like to perform?",
                                    O_ALL)
        self.operands = (Operand(0, self.operation), Operand(1, self.operation))
        self.wants_log = cps.Binary("Do you want the log (base 10) of the ratio?", False)
        self.final_multiplicand = cps.Float("What number would you like to multiply the result by?",1)
        self.final_exponent = cps.Float("What power would you like to raise the result to?",1)
            
    def settings(self):
        return ([self.output_feature_name, self.operation] +
                self.operands[0].settings() + self.operands[1].settings() + 
                [self.wants_log, self.final_multiplicand, self.final_exponent])

    def on_post_load(self, pipeline):
        '''Fixup any measurement names that might have been ambiguously loaded
        
        pipeline - for access to other module's measurements
        '''
        for operand in self.operands:
            measurement = operand.operand_measurement.value
            pieces = measurement.split('_')
            if len(pieces) == 4:
                measurement = pipeline.synthesize_measurement_name(self,
                                                                   operand.object,
                                                                   pieces[0],
                                                                   pieces[1],
                                                                   pieces[2],
                                                                   pieces[3])
                operand.operand_measurement.value = measurement
                 
    def visible_settings(self):
        return ([self.output_feature_name, self.operation] +
                self.operands[0].visible_settings() + 
                self.operands[1].visible_settings() + 
                [self.wants_log, self.final_multiplicand, self.final_exponent])
        

    def run(self, workspace):
        m = workspace.measurements
        values = []
        input_values = []
        has_image_measurement = any([operand.object == cpmeas.IMAGE
                                     for operand in self.operands])
        all_image_measurements = all([operand.object == cpmeas.IMAGE
                                     for operand in self.operands])
        all_object_names = list(set([operand.operand_objects.value
                                     for operand in self.operands
                                     if operand.object != cpmeas.IMAGE]))
        for operand in self.operands:
            value = m.get_current_measurement(operand.object,
                                              operand.operand_measurement.value)
            if isinstance(value, str) or isinstance(value, unicode):
                try:
                    value = float(value)
                except ValueError:
                    raise ValueError("Unable to use non-numeric value in measurement, %s"%operand.measurement.value)
            input_values.append(value)
            value *= operand.multiplicand.value
            value **= operand.exponent.value
            values.append(value)
        
        if (not has_image_measurement) and len(values[0]) != len(values[1]):
            raise ValueError("Incompatable objects: %s has %d objects and %s has %d objects"%
                             (operands[0].operand_objects.value, len(values[0]),
                              operands[1].operand_objects.value, len(values[1])))
        
        if self.operation == O_ADD:
            result = values[0]+values[1]
        elif self.operation == O_SUBTRACT:
            result = values[0]-values[1]
        elif self.operation == O_MULTIPLY:
            result = values[0] * values[1]
        elif self.operation == O_DIVIDE:
            result = values[0] / values[1]
        else:
            raise NotImplementedError("Unsupported operation: %s"%self.operation.value)
        #
        # Post-operation rescaling
        #
        if self.wants_log.value:
            result = np.log10(result)
        result *= self.final_multiplicand.value
        result **= self.final_exponent.value
        feature = self.measurement_name()
        if all_image_measurements:
            m.add_image_measurement(feature, result)
        else:
            for object_name in all_object_names:
                m.add_measurement(object_name, feature, result)

    def measurement_name(self):
        return "Math_"+self.output_feature_name.value
            
    def get_measurement_columns(self, pipeline):
        all_object_names = [operand.operand_objects.value
                            for operand in self.operands
                            if operand.object != cpmeas.IMAGE]
        if len(all_object_names):
            return [(name, self.measurement_name(), cpmeas.COLTYPE_FLOAT)
                    for name in all_object_names]
        else:
            return [(cpmeas.IMAGE, 
                     self.measurement_name(), 
                     cpmeas.COLTYPE_FLOAT)]

    def get_categories(self, pipeline, object_name):
        all_object_names = [operand.operand_objects.value
                            for operand in self.operands
                            if operand.object != cpmeas.IMAGE]
        if len(all_object_names):
            if object_name in all_object_names:
                return ["Math"]
        elif object_name == cpmeas.IMAGE:
            return ["Math"]
        return []

    def get_measurements(self, pipeline, object_name, category):
        if category in self.get_categories(pipeline, object_name):
            return [self.output_feature_name.value]
        return []
    
    def backwards_compatibilize(self, setting_values, variable_revision_number,
                                module_name, from_matlab):
        if from_matlab and variable_revision_number == 6:
            new_setting_values = [setting_values[16], # output feature name
                                  setting_values[15]] # operation
            for i,multiply_factor_idx in ((0,11),(5,12)):
                object_name = setting_values[i]
                category = setting_values[i+1]
                feature = setting_values[i+2]
                measurement_image = setting_values[i+3]
                scale = setting_values[i+4]
                measurement = category+'_'+feature
                if len(measurement_image):
                    measurement += '_' + measurement_image
                if len(scale):
                    measurement += '_' + scale
                object_choice = (MC_IMAGE if object_name == cpmeas.IMAGE
                                 else MC_OBJECT) 
                new_setting_values += [object_choice,
                                       object_name,
                                       measurement,
                                       setting_values[multiply_factor_idx],
                                       "1"] # exponent
            new_setting_values += [setting_values[10], # wants log
                                   setting_values[14], # final multiplier
                                   setting_values[13]] # final exponent
            setting_values = new_setting_values
            from_matlab = False
            variable_revision_number = 1
        return setting_values, variable_revision_number, from_matlab

